builder/slide_and_fold_storage.cc

```diff
@@ -264,13 +264,16 @@ class slide_and_fold : public node_mutator {
     // is produced there.
     if (loops.size() < 2 || !is_produced_by(output, body)) return;
 
-    auto ff = fold_factors[output];
     bool did_overlapped_fold = false;
 
-    if (ff) {
-      for (int d = 0; d < static_cast<int>(ff->size()); ++d) {
-        expr overlap = (*ff)[d].second;
-        did_overlapped_fold = did_overlapped_fold || overlap.defined();
+    if (fold_factors[output]) {
+      for (int d = 0; d < static_cast<int>(fold_factors[output]->size()); ++d) {
+        expr fold = (*fold_factors[output])[d].first;
+        expr overlap = (*fold_factors[output])[d].second;
+        if (!is_finite(fold)) continue;
+        // If fold is finite and bounds don't overlap the fold and overlap
+        // will be set to the same expr.
+        did_overlapped_fold = did_overlapped_fold || !fold.same_as(overlap);
```

Comment:
`same_as` is shallow equality, vs. `match(fold, overlap)` would be deep equality. Is shallow equality OK here? Should we just use `match(fold, overlap)` anyways to be safe? I don't think this is performance sensitive.

Reply:
I think same_as should be fine in this case, but changed to match just in case.